# Contents
### 1. Importing Libraries
### 2. Importing Data
### 3. Merge df_ords_prods_comb with df_prods

# 1. Importing Libraries

In [4]:
import pandas as pd
import numpy as np
import os

# 2. Importing Data

In [5]:
#set path

path = r'C:\Documents\Career Foundry\Python\Instacart Basket Analysis\02 Data'

In [6]:
# import pickled df

df_ords_prods_comb = pd.read_pickle(os.path.join(path, 'Prepped Data', 'orders_products_combined.pkl'))

In [7]:
df_ords_prods_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     0.0         196                  1          0   both  
1                     0.0       14084                  2          0   both  
2                     0.0       12427                  3          0   both  
3                     0.0       26088                  4          0   both  
4                     0.0       26405                  5          0   both

In [8]:
df_ords_prods_comb.shape

(32434489, 10)

In [9]:
# import Products csv

df_prods = pd.read_csv(os.path.join(path, 'Prepped Data', 'products_checked.csv'), index_col = False)

In [10]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [14]:
# drop extra column

df_prods.drop(df_prods.columns[[0]], axis = 1, inplace = True)

In [11]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [12]:
df_prods.shape

(49672, 6)

In [17]:
# Export df_prods with extra column dropped to csv

df_prods.to_csv(os.path.join(path, 'Prepped Data', 'Products_Checked.csv'))

In [13]:
df_ords_prods_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     0.0         196                  1          0   both  
1                     0.0       14084                  2          0   both  
2                     0.0       12427                  3          0   both  
3                     0.0       26088                  4          0   both  
4                     0.0       26405                  5          0   both

In [21]:
del df_m

In [22]:
import gc

In [23]:
gc.collect()

190

In [24]:
# drop indicator row from df_ords_prods_comb

df_ords_prods_comb.drop(columns=['_merge'], inplace=True)

In [26]:
df_ords_prods_comb.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     0.0         196                  1          0  
1                     0.0       14084                  2          0  
2                     0.0       12427                  3          0  
3                     0.0       26088                  4          0  
4                     0.0       26405                  5          0

In [29]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [37]:
# drop unnamed column from df_prods

df_prods.drop(df_prods.columns[[0]], axis = 1, inplace = True)

In [38]:
df_prods.head()

product_id                                       product_name  aisle_id  \
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   
5           6                                       Dry Nose Oil        11   

   department_id  prices  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3  
5             11     2.6

# 3. Merge df_ords_prods_comb with df_prods

In [40]:
# df_m merge df_ords_prods_comb with df_prods, indicator=false

df_m = df_ords_prods_comb.merge(df_prods, on = 'product_id', indicator = False)

In [41]:
df_m.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

In [42]:
df_m.shape

(32403007, 13)

In [43]:
df_ords_prods_comb.shape

(32434489, 9)

## df_m has less rows than df_ords_prods_comb which I think means that there was some info in df_ords_prods_comb that doesn't exist in df_prods and visa versa.

In [49]:
%whos DataFrame

Variable             Type         Data/Info
-------------------------------------------
df_m                 DataFrame              order_id  user_<...>403007 rows x 13 columns]
df_mo                DataFrame               order_id  user<...>435070 rows x 14 columns]
df_ords_prods_comb   DataFrame              order_id  user_<...>2434489 rows x 9 columns]
df_prods             DataFrame           product_id        <...>n[49671 rows x 5 columns]


In [47]:
# df_mo merge df_ords_prods_comb with df_prods, indicator=True, out=outer

df_mo = df_ords_prods_comb.merge(df_prods, on = 'product_id', how = 'outer', indicator = True)

In [52]:
df_mo['_merge'].value_counts()

both          32403007
left_only        32052
right_only          11
Name: _merge, dtype: int64

the count of 'both' in this df (merged with 'outer'), 32403007, is the same total count of df_m (merged with 'inner')
so df_ords_prods_comb has 32052 rows that df_prods doesn't have and df_prods has 11 rows that df_ords_prods_comb doesn't have.

I confirmed the results using the merge flag with df_mo (merged using 'outer') because counting the indicator column when merged with both is redundant - that is at least what the lesson stated and it makes sense to me.  However, the memory creating df_mo was crazy.  I had to leave the computer alone for approx 1/2 hour waiting for the * to clear.

In [53]:
del df_mo

In [58]:
%whos DataFrame

Variable             Type         Data/Info
-------------------------------------------
df_m                 DataFrame              order_id  user_<...>403007 rows x 13 columns]
df_ords_prods_comb   DataFrame              order_id  user_<...>2434489 rows x 9 columns]
df_prods             DataFrame           product_id        <...>n[49671 rows x 5 columns]


In [63]:
#export merged df as pkl to 'Prepped Data'

df_m.to_pickle(os.path.join(path, 'Prepped Data', 'orders_products_merged.pkl'))